libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1941,6 +1941,15 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 			ContractDefinition const& contract = contractType.contractDefinition();
 			define(_memberAccess) << formatNumber(u256{contract.interfaceId()} << (256 - 32)) << "\n";
 		}
+		else if (member == "typehash")
+		{
+			Type const* arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
+			solAssert(!arg);
```

Comment:
I think it needs to be:

```suggestion
			solAssert(arg);
```

since `solAssert(false)` would throw with `std::exception::what: Solidity assertion failed`.

Reply:
Now:

```bash
./build/solc/solc --ir ./test/libsolidity/smtCheckerTests/types/type_typehash.sol
```

compiles without errors but we have:

```solidity
/// @ast-id 99
/// @src 0:357:913  "function f() public pure {..."
function fun_f_99() {

    /// @src 0:393:410  "type(S1).typehash"
    let expr_47 := 0x78a822935e38445215ba7404686b919cbbef5725cbf9231d92802f542d7456e0
    /// @src 0:414:440  "keccak256(\"S1(uint256 x)\")"
    let expr_50 := 0x78a822935e38445215ba7404686b919cbbef5725cbf9231d92802f542d7456e0
    /// @src 0:393:440  "type(S1).typehash == keccak256(\"S1(uint256 x)\")"
    let expr_51 := eq(cleanup_t_bytes32(expr_47), cleanup_t_bytes32(expr_50))
    /// @src 0:386:441  "assert(type(S1).typehash == keccak256(\"S1(uint256 x)\"))"
    assert_helper(expr_51)
    /// @src 0:452:469  "type(S2).typehash"
    let expr_58 := 0x6c397ebd50462a81423e44830702ee1214cb9ab734bf173eb55f04238c9d398f
    /// @src 0:473:509  "keccak256(\"S2(uint256 x,address y)\")"
    let expr_61 := 0x6c397ebd50462a81423e44830702ee1214cb9ab734bf173eb55f04238c9d398f
    /// @src 0:452:509  "type(S2).typehash == keccak256(\"S2(uint256 x,address y)\")"
    let expr_62 := eq(cleanup_t_bytes32(expr_58), cleanup_t_bytes32(expr_61))
    /// @src 0:445:510  "assert(type(S2).typehash == keccak256(\"S2(uint256 x,address y)\"))"
    assert_helper(expr_62)
    /// @src 0:521:538  "type(S3).typehash"
    let expr_69 := 0x0d52677e9a6f5c217ec68867e6855a5221fe3d45c3334e588c88f0186968dda1
    /// @src 0:542:610  "keccak256(\"S3(uint256 x,address y,S2 third)S2(uint256 x,address y)\")"
    let expr_72 := 0xfa5685568fb2f15c09479ecbcfe9d0494743d804587d1966db67d5e62ea4344a
    /// @src 0:521:610  "type(S3).typehash == keccak256(\"S3(uint256 x,address y,S2 third)S2(uint256 x,address y)\")"
    let expr_73 := eq(cleanup_t_bytes32(expr_69), cleanup_t_bytes32(expr_72))
    /// @src 0:514:611  "assert(type(S3).typehash == keccak256(\"S3(uint256 x,address y,S2 third)S2(uint256 x,address y)\"))"
    assert_helper(expr_73)
    /// @src 0:628:645  "type(S4).typehash"
    let expr_80 := 0x5fc4fdd5ba2cd70a58869ff54ddc276bd05f4fc6b588ee53d67cacc1c29423e9
    /// @src 0:649:734  "keccak256(\"S4(S3 one,S2 two)S2(uint256 x,address y)S3(uint256 x,address y,S2 third)\")"
    let expr_83 := 0x17ed8da37c0446eeffef8cd38d116505b399b5644fdc3a59f8a68a68dd5d4178
    /// @src 0:628:734  "type(S4).typehash == keccak256(\"S4(S3 one,S2 two)S2(uint256 x,address y)S3(uint256 x,address y,S2 third)\")"
    let expr_84 := eq(cleanup_t_bytes32(expr_80), cleanup_t_bytes32(expr_83))
    /// @src 0:621:735  "assert(type(S4).typehash == keccak256(\"S4(S3 one,S2 two)S2(uint256 x,address y)S3(uint256 x,address y,S2 third)\"))"
    assert_helper(expr_84)
    /// @src 0:752:769  "type(S5).typehash"
    let expr_91 := 0x2ea2604a167f90524bcb3dd220a98d9eea77a989a8b16e8f9c35013dfd18a092
    /// @src 0:773:908  "keccak256(\"S5(S2 two,S1 one,S3 three,S4[5] four)S1(uint256 x)S2(uint256 x,address y)S3(uint256 x,address y,S2 third)S4(S3 one,S2 two)\")"
    let expr_94 := 0x5e52252fbbc0eda2d75f57c57d47fbec3bc6b215a9a3790c7f7ca44a36eb5185
    /// @src 0:752:908  "type(S5).typehash == keccak256(\"S5(S2 two,S1 one,S3 three,S4[5] four)S1(uint256 x)S2(uint256 x,address y)S3(uint256 x,address y,S2 third)S4(S3 one,S2 two)\")"
    let expr_95 := eq(cleanup_t_bytes32(expr_91), cleanup_t_bytes32(expr_94))
    /// @src 0:745:909  "assert(type(S5).typehash == keccak256(\"S5(S2 two,S1 one,S3 three,S4[5] four)S1(uint256 x)S2(uint256 x,address y)S3(uint256 x,address y,S2 third)S4(S3 one,S2 two)\"))"
    assert_helper(expr_95)

}
```

Reply:
some of these typehashes don't match because of the following [`eip712EncodeType()`](https://github.com/ethereum/solidity/blob/78d3cef5e26e887f02859f45a67dfd923e064bc9/libsolidity/ast/AST.cpp#L421):

```solidity
S3(uint256 x,I1 y,C.S2 third)S2(uint256 x,address y)
S4(C.S3 one,C.S2 two)S2(uint256 x,address y)S3(uint256 x,I1 y,C.S2 third)
S5(C.S2 two,C.S1 one,C.S3 three,C.S4[5] four)S1(uint256 x)S2(uint256 x,address y)S3(uint256 x,I1 y,C.S2 third)

Reply:
see:
- https://github.com/ethereum/solidity/pull/14167#discussion_r1193262629